### ISSUE TITLE
t.application.app.AppProfiler handles options oddly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3355]
|<img alt="colin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| colin reported|
|-|-|
|Trac ID|trac#3355|
|Type|defect|
|Created|2008-07-14 10:11:16Z|
|Branch|https://github.com/twisted/twisted/tree/default-profiler-3355|

If (like me..) you're a naughty boy and you have your own bootstrap layer which can be py2exe'd (because you work for evil Mega Corp) then the new AppProfiler breaks claiming None is an unsupported profiler name.

Since twistd and co pass this default down to AppProfiler, and since the options has a .get default of None I don't get why None isn't being handled as 'hotshot' (which is the real default). 

I don't know what to suggest, because the person who is looking after this beast needs to comment first :) 

A simple "fix" is just to change 'self.profiler = options.get("profiler", None)' to 'self.profiler = options.get("profiler", "hotshot")', but I'm doubtful that this is correct in any way.

<details><summary>Searchable metadata</summary>

```
trac-id__3355 3355
type__defect defect
reporter__colin colin
priority__lowest lowest
milestone__ 
branch__branches_default_profiler_3355 branches/default-profiler-3355
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1216030276000000 1216030276000000
changetime__1216888241000000 1216888241000000
version__None None
owner__ 
cc__colin@... cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @glyph|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Colin, can you be a bit more specific?  "your own bootstrap layer" reads to me as "a pile of arbitrary code".  Can you provide a working example of a miniature version of this "bootstrap layer" that doesn't work?  The fact that you labeled this a "regression" implies that it used to work, but `AppProfiler` wasn't in 2.5 so I'm not sure what you mean.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

It's a bit weird that an `AppProfiler` is _always_ created, even if the configuration blob doesn't indicate that profiling will be performed.  Later on, another layer decides that the config doesn't require profiling and then ignores the profiler object.  It'd probably simplify the code and the interface if the profiler were only created if profiling was demanded by the configuration.  Making this so while maintaining API compatibility might require some creativity.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [24291]) Branching to 'default-profiler-3355'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [24292]) Default to the hotshot profiler, test it.

Refs [#3355](https://github.com/twisted/twisted/issues/3355)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

Defaulting to the hotshot profiler makes sense. I think changing the rest too much doesn't worth it for now, especially for this problem. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [24293]) Fix cProfile management.

Refs [#3355](https://github.com/twisted/twisted/issues/3355)
Refs [#3356](https://github.com/twisted/twisted/issues/3356)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

I took the opportunity to fix [#3356](https://github.com/twisted/twisted/issues/3356) here too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

For the `--help` output, I'd suggest dropping _between_ and just putting the list into parenthesis (`imaginary.language.itemizedStringList` would be great here ;).

Otherwise, looks great, merge when you're happy with it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

btw, build results - http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/default-profiler-3355

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [24330]) Merge default-profiler-3355

Author: therve
Reviewer: exarkun
Fixes [#3355](https://github.com/twisted/twisted/issues/3355)
Fixes [#3356](https://github.com/twisted/twisted/issues/3356)

Make hotshot explicitely the default profiler in AppProfiler, and add cProfile
to twistd --help.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
